src/main/scala/geekie/mapred/MapredWorker.scala

```diff
@@ -0,0 +1,83 @@
+package geekie.mapred
+
+import akka.actor.{Actor, ActorRef, Props}
+import akka.routing.{ActorRefRoutee, RemoveRoutee}
+import geekie.mapred.MapredWorker._
+
+import scala.collection.TraversableLike
+import scala.concurrent.duration._
+import scala.language.postfixOps
+import scala.reflect.ClassTag
+
+class MapredWorker[A: ClassTag, K: ClassTag, V: ClassTag](mf: A => TraversableOnce[MapredWorker.KeyVal[K, V]],
+                                                          rf: (V, V) => V,
+                                                          ff: Map[K, V] => Unit) extends Actor {
+
+  var acc = Map.empty[K, V]
+
+  def aggregateNewData(kvs: TraversableOnce[KeyVal[K, V]]) = kvs foreach {
+    case KeyVal(k, v) =>
+      acc += (k -> (if (acc contains k) rf(acc(k), v) else v))
```

Comment:
Usually the FP patrol dictates you should try to work exclusively with immutables, then a var of an immutable container if you need to, and then mutable... I have replaced the current code with the proper "Akka way" of representing state: as a parameter of the receive function, and with a `context.become(...)` at every case where the state is updated.

Who knows if a mutable map might give some speed improvement. My bet is it would be minor, but we can try someday... But remember we are also interested in safety here, with immutables we can safely share the map between actors, and we also avoid ever having to copy the whole structure to preserve it. This application here is not so complex, so it is probably safe to use a mutable map (we never share the map actually). But otherwise we would be inviting the evil of shared mutable state into our program!...